src/app.js

```diff
@@ -221,30 +216,128 @@ export default function (app) {
 
   function patch(parent, element, oldNode, node) {
     if (oldNode == null) {
-      element = parent.appendChild(createElementFrom(node))
+      element = parent.insertBefore(createElementFrom(node), element)
 
-    } else if (node == null) {
-      batch.push(function () {
-        if (oldNode.data && oldNode.data.onRemove) {
-          oldNode.data.onRemove(element)
-        }
-        parent.removeChild(element)
-      })
     } else if (node.tag && node.tag === oldNode.tag) {
-      updateElementData(element, node.data, oldNode.data)
+      updateElementData(element, oldNode.data, node.data)
 
       var len = node.children.length
       var oldLen = oldNode.children.length
+      var reusableChildren = {}
+      var oldElements = []
+      var newKeys = {}
+
+
+      for (var i = 0; i < oldLen; i++) {
+        var oldElement = element.childNodes[i]
+        oldElements[i] = oldElement
 
-      for (var i = 0; i < len || i < oldLen; i++) {
-        patch(element, element.childNodes[i], oldNode.children[i], node.children[i])
+        var oldChild = oldNode.children[i]
+        var oldKey = getKeyFrom(oldChild)
+
+        if (null != oldKey) {
+          reusableChildren[oldKey] = [oldElement, oldChild]
+        }
       }
 
+      var i = 0
+      var j = 0
+
+      while (j < len) {
+        var oldElement = oldElements[i]
+        var oldChild = oldNode.children[i]
+        var newChild = node.children[j]
+
+        var oldKey = getKeyFrom(oldChild)
+        if (newKeys[oldKey]) {
+          i++
+          continue
+        }
+
+        var newKey = getKeyFrom(newChild)
+
+        var reusableChild = reusableChildren[newKey]
+        var reusableElement = 0
+        var reusableNode = 0
+
+        if (reusableChild) {
+          reusableElement = reusableChild[0]
+          reusableNode = reusableChild[1]
+        }
+
+        if (null == oldKey && null == newKey) {
+          patch(element, oldElement, oldChild, newChild)
+          j++
+          i++
+
+        } else if (null == oldKey && null != newKey) {
+          if (reusableElement) {
+            element.insertBefore(reusableElement, oldElement)
+            patch(element, reusableElement, reusableNode, newChild)
+          } else {
+            patch(element, oldElement, null, newChild)
+          }
```

Comment:
Yes makes sense, except it's a little more specific (I figured out after looking a little closer at the algo): It's when we encounter a child in the new tree *with a key that was not used in the old tree*

It makes sense those lines wouldn't be covered then because I have no tests where the new tree has keys not present in the old.

So all is well... nothing to do here :)